codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/UnionGenerator.java

```diff
@@ -51,48 +52,41 @@ public void run() {
             writer.write("is$L()", symbol.getName());
         }).write("");
 
-        // Create type aliases for each member that satisfy the interface.
+        // Create structs for each member that satisfy the interface.
         for (MemberShape member : shape.getAllMembers().values()) {
             Symbol memberSymbol = symbolProvider.toSymbol(member);
             String exportedMemberName = symbol.getName() + symbolProvider.toMemberName(member);
-            String unExportedMemberName = StringUtils.uncapitalize(exportedMemberName);
+            Shape target = model.expectShape(member.getTarget());
 
-            // Create an interface for the member
+            // Create the member's concrete type
             writer.writeMemberDocs(model, member);
-            writer.openBlock("type $L interface {", "}", exportedMemberName, () -> {
-                writer.write("$L", symbol.getName());
-                writer.write("is$L()", exportedMemberName);
-                writer.write("Value() $T", memberSymbol);
+            writer.openBlock("type $L struct {", "}", exportedMemberName, () -> {
+                // Union members can't have null values, so for simple shapes we don't
+                // use pointers. We have to use pointers for complex shapes since,
+                // for example, we could still have a map that's empty or which has
+                // null values.
+                if (target instanceof SimpleShape) {
+                    writer.write("Value $T", memberSymbol);
+                } else {
+                    writer.write("Value $P", memberSymbol);
+                }
             });
 
-            writer.openBlock("func As$L(v $T) $L {", "}", exportedMemberName, memberSymbol, exportedMemberName, () -> {
-                writer.write("return $L{value: v}", unExportedMemberName);
-            });
-
-            // Create the member's un-exported concrete type
-            writer.openBlock("type $L struct {", "}", unExportedMemberName, () -> {
-                writer.write("value $T", memberSymbol);
-            });
-
-            writer.write("func ($L) is$L() {}", unExportedMemberName, symbol.getName());
-            writer.write("func ($L) is$L() {}", unExportedMemberName, exportedMemberName);
-            writer.openBlock("func (v $L) Value() $T {", "}", unExportedMemberName, memberSymbol, () -> {
-                writer.write("return v.value");
-            });
+            writer.write("func ($L) is$L() {}", exportedMemberName, symbol.getName());
```

Comment:
We probably want to do a pointer receiver here for this method instead of value. Pointer receiver ensures that when user does type assert for this union member wrapper, they don't need to compare against both value and pointer types. This would prevents the guess work, guessing if the interface value was a pointer or value.

I think typed errors used this pattern of pointer receivers as well.